core/src/torii/routing.rs

```diff
@@ -167,10 +167,21 @@ async fn handle_subscription(events: EventsSender, stream: WebSocket) -> eyre::R
     let mut consumer = Consumer::new(stream).await?;
 
     loop {
-        let event = events.recv().await?;
-
-        iroha_logger::trace!(?event);
-        consumer.consume(event).await?;
+        tokio::select! {
+            message_opt = consumer.stream_mut().next() => {
```

Comment:
Yes, it consumes new message from stream.

Berfore that PR server was listening only for `EventReceived` messages from client and only if an event has occurred. If client was sending something berfore an event happened, server wasn't listening for that. So I assume that by design only one client behaviour was correct -- to send `EventReceived` when a new event was received.

Now I had to implement possibility to close connection cleanly. That's a new type of message (may be it should be presented on iroha message structures?), that can be created with `ws::Message::close()`. That message can be sent at any time during subscription, so I'm listening for that type of mesages here. There is no other types of messages right now.

Maybe it will be good to log if it's not `Close` message was received?

Reply:
I think this code is all right.
The 1st branch is listening to external stream message.
The 2nd branch is listening to internal events

Reply:
ok, I'm just pointing out what I think might be an issue. I'm not always correct. If you think all is good I trust you

Reply:
but what happens if message other than close message is received here? it will be swept under the carpet. Isn't there some peek operation on the stream?

Reply:
`warp` WebSocket does not support peeking. So if we are gonna implement it ourselves on `Stream` trait, then what will be the difference if we still has to use the `next()` method?

Reply:
What I meant here is first peek then take out of stream, but since you have concurrency issues as commented below this will not help. After you resolve concurrency issue this may become a non issue or we'll have to figure out a way how to first peek and then take out of stream.